# Data Science – Упражнения для обработки текста


## NLP

### Упражнение 1: Часто встречающиеся слова
Найдите в «Моби Дике» наиболее часто используемые слова, которые не являются стоп-словами и знаками препинания. Подсказка: здесь может пригодиться [`str.isalpha()`](https://docs.python.org/3/library/stdtypes.html#str.isalpha).

In [8]:
import nltk
nltk.download('all')
from nltk.corpus import stopwords
stopwords = nltk.corpus.stopwords.words('english')
from nltk.book import *

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package basque_grammars is already up-to-date!
[nltk_data]    | Downloading package bcp47 to /root/nltk_data...
[nltk_data]    |   Package bcp47 is already up-to-dat

In [13]:
import numpy as np
from sklearn.neighbors import NearestNeighbors
from sklearn.neighbors import KNeighborsClassifier
from sklearn import svm
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
import string
from nltk.corpus import stopwords
from collections import Counter
from nltk.tokenize import word_tokenize
from nltk import FreqDist
from nltk.corpus import stopwords
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams['figure.figsize'] = (15, 9)
plt.style.use('ggplot')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [16]:
print("Всего слов:", len(text1))
print("Количество уникальных слов: ", len(set(text1)))

Всего слов: 260819
Количество уникальных слов:  19317


In [14]:
text = nltk.Text(text1) #создание объекта из text1

word_list = [word.lower() for word in text.tokens if word.isalpha() # Приведение слов в нижний регистр
  and word.lower() not in stop_words] #проверка слов в нижнем регистре, что они не входят в список стоп слов

In [18]:
print("Количество слов в обработанном списке:", len(word_list))
print("Количество уникальных слов в обработанном списке:", len(set(word_list)))

Количество слов в обработанном списке: 110459
Количество уникальных слов в обработанном списке: 16802


После обработки исходного текста от чисел и знаков препинания количество слов уменьшилось на 150360, а количество уникальных слов уменьшилось на 2515

---



In [15]:
fdist = FreqDist(word_list)# подсчёт частоты встречаемости слов в списке
most_common_words = fdist.most_common(10) # извлечение 50 часто встречающихся слов

for word, frequency in most_common_words: # цикл проходит по связrе слово-частота
  print(f"Слово |{word}| встречается {frequency} раз")#Вывод результата

Слово |whale| встречается 1226 раз
Слово |one| встречается 921 раз
Слово |like| встречается 647 раз
Слово |upon| встречается 566 раз
Слово |man| встречается 527 раз
Слово |ship| встречается 518 раз
Слово |ahab| встречается 511 раз
Слово |ye| встречается 472 раз
Слово |sea| встречается 455 раз
Слово |old| встречается 450 раз


Чаще всего встречается слово - "кит" с частотой 1226.

Слово "один" встречается 921 - это второе слово по частоте употребления.

Слово "как" встречается в тексте 647 раз.

In [19]:
text1.concordance("whale")#просмотр рядом стоящих слов с искомым словом "Кит"

Displaying 25 of 1226 matches:
s , and to teach them by what name a whale - fish is to be called in our tongue
t which is not true ." -- HACKLUYT " WHALE . ... Sw . and Dan . HVAL . This ani
ulted ." -- WEBSTER ' S DICTIONARY " WHALE . ... It is more immediately from th
ISH . WAL , DUTCH . HWAL , SWEDISH . WHALE , ICELANDIC . WHALE , ENGLISH . BALE
HWAL , SWEDISH . WHALE , ICELANDIC . WHALE , ENGLISH . BALEINE , FRENCH . BALLE
least , take the higgledy - piggledy whale statements , however authentic , in 
 dreadful gulf of this monster ' s ( whale ' s ) mouth , are immediately lost a
 patient Job ." -- RABELAIS . " This whale ' s liver was two cartloads ." -- ST
 Touching that monstrous bulk of the whale or ork we have received nothing cert
 of oil will be extracted out of one whale ." -- IBID . " HISTORY OF LIFE AND D
ise ." -- KING HENRY . " Very like a whale ." -- HAMLET . " Which to secure , n
restless paine , Like as the wounded whale to shore flies thro ' the maine ." -
. OF SPER

In [20]:
text1.concordance("one")

Displaying 25 of 921 matches:
 be too rosy - strong ; but with whom one sometimes loves to sit , and feel poo
an maketh a path to shine after him ; One would think the deep to be hoary ." -
e sea , appeared . Among the former , one was of a most monstrous size . ... Th
ifty yards long . He said that he was one of six who had killed sixty in two da
ntity of oil will be extracted out of one whale ." -- IBID . " HISTORY OF LIFE 
barrel of herrings in his belly . ... One of our harpooneers told me that he ca
 upon this coast ( Fife ) Anno 1652 , one eighty feet in length of the whale - 
seas , as I may say , by a hundred to one ; than we have to the northward of us
SSECTION OF A WHALE . ( A SMALL SIZED ONE .) " The aorta of a whale is larger i
g and sporting with each other , when one observed : there -- pointing to the s
1821 . " A mariner sat in the shrouds one night , The wind was piping free ; No
 boats engaged in the capture of this one whale , amounted altogether to 10 , 4
 ON WHALIN

In [21]:
text1.concordance("like")

Displaying 25 of 647 matches:
athan that maketh the seas to seethe like boiling pan ." -- LORD BACON ' S VER
ard bruise ." -- KING HENRY . " Very like a whale ." -- HAMLET . " Which to se
east , had bred his restless paine , Like as the wounded whale to shore flies 
ERMA CETI WHALE . VIDE HIS V . E . " Like Spencer ' s Talus with his modern fl
ng creatures , in the deep Stretched like a promontory sleeps or swims , And s
e fishes , you would make them speak like great wales ." -- GOLDSMITH TO JOHNS
ave : Gather ' d in shoals immense , like floating islands , Led by mysterious
s tail in the air , which , cracking like a whip , resounds to the distance of
rthward . What do you see ?-- Posted like silent sentinels all around the town
is pine - tree shakes down its sighs like leaves upon this shepherd ' s head ,
inction of such offices to those who like them . For my part , I abominate all
and make me jump from spar to spar , like a grasshopper in a May meadow . And 
 of the bill must have

## Упражнение 2

Вы злой спамер, который заметил, что многие люди пытаются запутать свою электронную почту, используя пометку: «alex at utah dot edu». Ниже приведены три примера текста таких писем. Попробуйте извлечь «alex at utah dot edu» и т. д. Начните с первой строки. Затем расширьте свое регулярное выражение, чтобы оно работало со всеми ними одновременно. Обратите внимание, что второе и третье сделать намного сложнее!

You're an evil Spammer who's observed that many people try to obfuscate their e-mail using this notation: "`alex at utah dot edu`". Below are three examples of such e-mails text. Try to extract "alex at utah dot edu", etc. Start with the first string. Then extend your regular expression to work on all of them at the same time. Note that the second and third are slightly harder to do!

1. \b - это "граница слова"

2. \w - это шаблон, который соответствует буквам, цифрам и подчеркиванию.

3. \s* - это шаблон, который соответствует любому количеству пробельных символов (пробел, табуляция, новая строка).

4. dot - это буквальная строка "dot".

5. (?: ... ) - это не захватывающая группа. Это позволяет группировать выражения без их сохранения в результирующем совпадении.

6. '* - это квантификатор, указывающий, что предыдущий элемент (в данном случае группа) может встречаться от нуля до бесконечности раз.

In [27]:
import re
html_smart = "You can reach me: alex at utah dot edu"
html_smart2 = "You can reach me: alex dot lex at utah dot edu"
html_smart3 = "You can reach me: alex dot lex at sci dot utah dot edu"

In [28]:
def testRegex(regex):
    for html in (html_smart, html_smart2, html_smart3):
        print(re.search(regex, html).group())

In [29]:
# TODO write your regex here
mail_regex =  r"[\w.]+(?: at | dot )[\w.]+(?: dot [\w.]+)?(?: at | dot )([\w.]+(?: dot [\w.]+)?)"

In [30]:
testRegex(mail_regex)

alex at utah dot edu
alex dot lex at utah dot edu
alex dot lex at sci dot utah


## Упражнение 2.2: Найдите наречия

Напишите регулярное выражение, которое находит все наречия в предложении. Наречия характеризуются окончанием на «ly».

In [31]:
text = "He was carefully disguised but captured quickly by police."

In [32]:
adverbs = re.findall(r'\b\w+ly\b', text)
for adverb in adverbs:
  print("Наречие -> ", adverb)

Наречие ->  carefully
Наречие ->  quickly


### Упражнение 2.3: Телефонные номера

Извлеките из текста номера телефонов, которые следуют шаблону (xxx) xxx-xxxx:

In [33]:
phone_numbers = "(857) 131-2235, (801) 134-2215, but this one (12) 13044441 shouldnt match. Also, this is common in twelve (12) countries and one (1) state"


\d{k} - обозначает k цифр подряд

In [16]:
phones = re.findall(r'\(\d{3}\) \d{3}-\d{4}', phone_numbers)
for phone in phones:
  print("Тел -> ", phone)

Тел ->  (857) 131-2235
Тел ->  (801) 134-2215


### Упражнение 2.4. HTML-контент

Извлеките содержимое между тегами `<b>` и `<i>`, но не другими тегами:

In [17]:
html_tags = "This is <b>important</b> and <u>very</u><i>timely</i>"

In [35]:
matches = re.findall(r'<b>(.*?)<\/b>|<i>(.*?)<\/i>', html_tags)

flattened_matches = [match for group in matches for match in group if match]
for i in flattened_matches:
  print(i)

important
timely
